### ISSUE TITLE
BTRFS snapshots need to be deleted twice for successful deletion

### ISSUE THREAD

**Issue by fxzzi** [linuxmint/timeshift#291]
**Describe the bug**
For some reason, deleting a snapshot in timeshift (I am using the GUI) will result in a "Snapshot deleted with errors". The snapshot stays visible in the GUI. However, if I just delete the snapshot again, it seems to successfully remove it.

**To Reproduce**
1. Create a manual snapshot (this happens with all snapshots, this is just to repro)
2. Attempt to delete the snapshot and observe the error
3. Attempt to delete it again and observe it deletes successfully

**Expected behavior**
The snapshot should be deleted successfully the first time.

**Screenshots**
![Screenshot from 24 04 02 11:45:52](https://github.com/linuxmint/timeshift/assets/18248986/6984790b-b1eb-4369-add2-3d101c9065e3)
![Screenshot from 24 04 02 11:46:04](https://github.com/linuxmint/timeshift/assets/18248986/08ecf2b9-91c4-4a2d-80d7-db11f8ae8a09)
![Screenshot from 24 04 02 11:46:07](https://github.com/linuxmint/timeshift/assets/18248986/afa87dca-92f9-4472-9a41-2b3f766e28a5)

Same behaviour occurring in cli: 
![image](https://github.com/linuxmint/timeshift/assets/18248986/56271d9d-0f9a-4a05-8fb3-240121bd2f27)

**System:**
 - Arch Linux
 - Hyprland Desktop
 - Timeshift 24.01.1

**Comment by Costasas** [comment]
Pop!_OS 22.04 LTS 6.8.0-76060800daily20240311-generic
 with BTRFS
Same error

**Comment by fxzzi** [comment]
Bump, is anyone able to help here? Is there something wrong with my fs? I've tried btrfs checks.

**Comment by atiredman** [comment]
Just wanted to chime in and say I am also seeing this issue with Btrfs Arch Linux (6.9.3) running the Cinnamon desktop (6.0.4). In my case, it can take 2 or even 3 attempts to delete some snapshots. 

I did some investigating by clicking Browse Files between deletion attempts. My snapshots typically contain 3 files/folders @, @home, and info.json. When I attempt to delete a snapshot with those three files/folders, the following occurs:

The first attempt results in the same “Snapshot(s) Deleted With Error” message Fxzzi shows, but browsing the files shows @home has been deleted.

The second attempt at deleting the snapshot results in the same “Snapshot(s) Deleted With Error” message, but this time @ is also deleted, leaving only info.json. 

The 3rd attempt finally deletes the snapshot successfully.

**Comment by fxzzi** [comment]
> The 3rd attempt finally deletes the snapshot successfully.

Probably takes you 3 attempts because you are backing up your @home also. For me it'll only take 2 as I only backup my main @ subvolume.

**Comment by atiredman** [comment]
> Probably takes you 3 attempts because you are backing up your @home also. For me it'll only take 2 as I only backup my main @ subvolume.

That is correct, if I don't back up my @home, it only takes 2 attempts to delete.

**Comment by fxzzi** [comment]
Bumping this. I still have no idea why this is happening.

**Comment by JGKle** [comment]
Seeing this too, on KDE Neon. It takes 3 tries to delete every snapshot:

* The first time I delete, the item remains in the Timeshift UI, and manually browsing to the files I can see that it removed `@home` only (the `@` folder and `info.json` remain).
* The second time I delete, the item remains in the Timeshift UI, now showing a usage of "0 B". Browsing to the files, `@` has been removed & only `info.json` remains.
* The third time I delete, it finally actually gets removed from the list & deleted.

Seems like this is a legitimate bug affecting a variety of distros. Even if there's no quick fix, would be nice if at least it wasn't ignored...?

**Comment by Kansai53** [comment]
Seeing this bug too. I have to delete every snapshot 3 times to get it to actually delete.

**Comment by Ricent82** [comment]
This is really agonizing. Are the developers going to even chime in about what's going on here?

**Comment by mtwebster** [comment]
Developer chiming in, I looked into this about a month ago, and was not able to reproduce it (though I have encountered it in the past). I'll give it another go when I get a chance.

**Comment by Ricent82** [comment]
Great, thanks for the quick reply. Strange tho - I have 2 different systems, and on both, it happens 100% of the time 🤔

**Comment by eloj** [comment]
> Developer chiming in, I looked into this about a month ago, and was not able to reproduce it (though I have encountered it in the past). I'll give it another go when I get a chance.

I too experience this issue. In the grand scheme, not the biggest deal, but feel free to hit me up if you need any system/config info. It's all very vanilla. I'm not including the \@home subvolume in the snapshots and I have to delete twice.

```
$ sudo timeshift --list-devices
Mounted '/dev/nvme0n1p2' at '/run/timeshift/64307/backup'
Mounted '/dev/nvme0n1p3' at '/run/timeshift/64307/backup-home'
```

**Comment by JimPix1** [comment]
I started having this problem today after I had done some messing with my timeshift partitions on a live iso, can't seem to resolve it either. It now shows sizes for the "Sizes" and "Unshared" sections which it previously did not. 
![image](https://github.com/user-attachments/assets/cde9a949-f017-46ef-ab97-39892d11ddff)

**Comment by fxzzi** [comment]
hey all, have a try after running `sudo btrfs quota disable /`

the size and unshared sizes will disappear, but the snapshots seem to delete correctly for me now first try.

**Comment by Ricent82** [comment]
Confirmed, the issue doesn't occur after running that command.

Just to clarify: hopefully this is just a test/debugging step, & not final solution? Since knowing the size of the snapshots is pretty important

**Comment by fxzzi** [comment]
> Confirmed, the issue doesn't occur after running that command.
> 
> Just to clarify: hopefully this is just a test/debugging step, & not final solution? Since knowing the size of the snapshots is pretty important

Well we now know that the issue is related to qgroups and timeshift not handling them in a fully correct manner I guess.

Timeshift should correctly delete the subvols and then also delete the stale qgroups which are left.

**Comment by Ricent82** [comment]
Great! So just more info/stepping stone on the way to a solution :)

**Comment by JGKle** [comment]
Any hope of some motion here? The workaround is really sort of just trading one issue for another, because if you disable that, there's no way to know which snapshots are using up all your storage. So I end up doing a 3-step delete anyway: disable that, delete, re-enable that. No matter how you cut it it takes 3 steps to delete something.

**Comment by LaMarMot** [comment]
I just realized it is the same issue than https://github.com/linuxmint/timeshift/issues/287.

So, still no real workaround I guess?

Cheers!

**Comment by mtwebster** [comment]
I should have a fix for this in the next day or so - as hinted at above it's an issue with deletion of qgroups. The backup is deleted fine but the 'qgroup' deletion fails due to a device/resource busy issue. I'll probably disable this entirely, as there's a nice command (that works) to automatically remove 'stale' qgroups, that can be run at startup instead.

**Comment by LaMarMot** [comment]
Hey @mtwebster thanks for your answer :)

> The backup is deleted fine 

Have a look https://github.com/linuxmint/timeshift/issues/287 as I don't think the backup deletes correctly. I think it does so only the next time TimeShift takes a new snapshot (it then erase the one it fail last time, and fail to erase the one it's suppose to erase this time... but will erase it next time, failing the other one... etc... you get the idea)  

Well I'm not with any of my machines right now, I can't test that in real time, but it is what I understand from the logs of https://github.com/linuxmint/timeshift/issues/287 (but I might as well be totally wrong)

Thanks again, so much appreciated !

Cheers !
